librad/src/meta/identity.rs

```diff
@@ -0,0 +1,1091 @@
+        } else {
+            Err(Error::NoCurrentQuorum)
+        }
+    }
+}
+
+impl Identity<Quorum> {
+    /// Check that a document update is legal
+    /// (the cache is updated accordingly)
+    pub fn check_update(
+        self,
+        previous: Option<&Identity<Verified>>,
+        cache: &mut impl VerificationCache,
+    ) -> Result<Identity<Verified>, Error> {
+        if cache.is_verified(self.revision()) {
+            return Ok(self.with_status());
+        }
+
+        let result = match previous {
+            Some(previous) => match self.doc().replaces() {
+                Some(replaces) => {
+                    if self.root() != previous.root() {
+                        Err(Error::RootMismatch)
+                    } else if replaces != previous.revision() {
+                        Err(Error::PreviousLinkMismatch)
+                    } else if !previous.doc().delegations().check_quorum(self.signatures()) {
+                        Err(Error::NoCurrentQuorum)
+                    } else {
+                        Ok(self.with_status())
+                    }
+                },
+                None => Err(Error::PreviousLinkMismatch),
+            },
+            None => {
+                if self.doc().is_root() {
+                    Ok(self.with_status())
+                } else {
+                    Err(Error::PreviousLinkMismatch)
+                }
+            },
+        };
+
+        if let Ok(id) = &result {
+            cache.register_verified(id)?;
+        }
+
+        result
+    }
+}
+
+/// Signature commit trailer name
+const RAD_SIGNATURE_TRAILER_NAME: &str = "x-rad-signature";
+
+/// Append a signature as a commit trailer
+fn append_signatures(buf: &mut String, sigs: &BTreeMap<PublicKey, Signature>) {
+    buf.push_str("\n");
+    for (k, s) in sigs {
+        buf.push_str(&format!(
+            "{}: {} {}\n",
+            RAD_SIGNATURE_TRAILER_NAME,
+            k.to_bs58(),
+            s.to_bs58()
+        ));
+    }
+}
+
+/// Matches a signature as a commit trailer in a text line
+fn match_signature(line: &str) -> Option<(PublicKey, Signature)> {
+    let mut tokens = line
+        .strip_prefix(RAD_SIGNATURE_TRAILER_NAME)
+        .and_then(|line| line.strip_prefix(": "))?
+        .split(' ');
+
+    let key_text = tokens.next()?;
+    let sig_text = tokens.next()?;
+    if tokens.next().is_some() {
+        return None;
+    }
+
+    Some((
+        PublicKey::from_bs58(key_text)?,
+        Signature::from_bs58(sig_text)?,
+    ))
+}
+
+/// Extract all signatures from a commit message
+pub fn parse_signatures(buf: Option<&str>) -> BTreeMap<PublicKey, Signature> {
+    let mut sigs = BTreeMap::new();
+    if let Some(buf) = buf {
+        for line in buf.split('\n') {
+            if let Some((k, s)) = match_signature(line) {
+                sigs.insert(k, s);
+            }
+        }
+    }
+    sigs
+}
+
+/// Identity API on top of a git repository
+pub struct IdentityStore<'a> {
```

Comment:
Can we split out the git related things into different modules? It would help to separate boundaries.